File: backend/service/src/main/java/com/star/models/energyamount/EnergyAmount.java
package com.star.models.energyamount;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.star.models.common.ValidationRegex;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import javax.validation.constraints.NotBlank;
import javax.validation.constraints.Pattern;

/**
 * Copyright (c) 2022, Enedis (https://www.enedis.fr), RTE (http://www.rte-france.com)
 * SPDX-License-Identifier: Apache-2.0
 */
@Builder
@Data
@NoArgsConstructor
@AllArgsConstructor
public class EnergyAmount {

    @JsonIgnore
    private String docType;

    private String energyAmountMarketDocumentMrid;

    private String activationDocumentMrid;

    private String registeredResourceMrid;

    @NotBlank(message = "Le champ quantity est obligatoire")
    private String quantity;

    @NotBlank(message = "Le champ measurementUnitName est obligatoire")
    private String measurementUnitName;

    private String revisionNumber;

    private String businessType;

    private String docStatus;

    private String processType;

    private String classificationType;

    @NotBlank(message = "Le champ areaDomain est obligatoire")
    private String areaDomain;

    @NotBlank(message = "Le champ senderMarketParticipantMrid est obligatoire")
    private String senderMarketParticipantMrid;

    @NotBlank(message = "Le champ senderMarketParticipantRole est obligatoire")
    private String senderMarketParticipantRole;

    @NotBlank(message = "Le champ receiverMarketParticipantMrid est obligatoire")
    private String receiverMarketParticipantMrid;

    @NotBlank(message = "Le champ receiverMarketParticipantRole est obligatoire")
    private String receiverMarketParticipantRole;

    @Pattern(regexp = ValidationRegex.DATETIME_OR_EMPTY_REGEX, message = "Le champ createdDateTime doit être au format : " + ValidationRegex.DATETIME_REGEX_STR)
    private String createdDateTime;

    @NotBlank(message = "Le champ timeInterval est obligatoire")
    private String timeInterval;

}

File: chaincode/src/model/energyAmount.ts
/*
 * SPDX-License-Identifier: Apache-2.0
 */
import * as Yup from 'yup';
import { DataActionType } from '../enums/DataActionType';

export class EnergyAmount {
    public static formatString(inputString: string) : EnergyAmount {
        let energyAmountObj: EnergyAmount;
        try {
            energyAmountObj = JSON.parse(inputString);
        } catch (error) {
            throw new Error(`ERROR EnergyAmount-> Input string NON-JSON value`);
        }

        try {
            EnergyAmount.schema.validateSync(
                energyAmountObj,
                {strict: true, abortEarly: false},
            );
        } catch (error) {
            throw error;
        }
        return energyAmountObj;
    }

    public static formatListString(inputString: string) : EnergyAmount[] {
        let energyAmountList: EnergyAmount[] = [];
        try {
            energyAmountList = JSON.parse(inputString);
        } catch (error) {
            throw new Error(`ERROR EnergyAmount by list-> Input string NON-JSON value`);
        }

        if (energyAmountList && energyAmountList.length > 0) {
            for (var energyAmountObj of energyAmountList) {
                try {
                    EnergyAmount.schema.validateSync(
                        energyAmountObj,
                        {strict: true, abortEarly: false},
                    );
                } catch (error) {
                    throw error;
                }
            }
        }
        return energyAmountList;
    }

    public static readonly schema = Yup.object().shape({
        activationDocumentMrid: Yup.string().required(
            'activationDocumentMrid is a compulsory string').typeError('activationDocumentMrid is a compulsory string',
        ),
        areaDomain: Yup.string().required('areaDomain is a compulsory string').typeError('areaDomain is a compulsory string'),
        businessType: Yup.string().notRequired(),
        classificationType: Yup.string().notRequired(),
        createdDateTime: Yup.string().required('createdDateTime is a compulsory string'),
        docStatus: Yup.string().notRequired(),
        docType: Yup.string().notRequired(),
        energyAmountMarketDocumentMrid: Yup.string().required(
            'energyAmountMarketDocumentMrid is a compulsory string',
        ),
        measurementUnitName: Yup.string().required('measurementUnitName is a compulsory string'),
        processType: Yup.string().notRequired(),
        quantity: Yup.string().required(
            'quantity is a compulsory string').typeError('quantity is a compulsory string',
        ),
        receiverMarketParticipantMrid: Yup.string().required(
            'receiverMarketParticipantMrid is a compulsory string').typeError('receiverMarketParticipantMrid must be a string',
        ),
        receiverMarketParticipantRole: Yup.string().required(
            'receiverMarketParticipantRole is a compulsory string').typeError('receiverMarketParticipantRole must be a string',
        ),
        registeredResourceMrid: Yup.string().notRequired(),
        revisionNumber: Yup.string().notRequired(),
        senderMarketParticipantMrid: Yup.string().required(
            'senderMarketParticipantMrid is a compulsory string').typeError('senderMarketParticipantMrid must be a string',
        ),
        senderMarketParticipantRole: Yup.string().required(
            'senderMarketParticipantRole is a compulsory string',
        ),
        timeInterval: Yup.string().required('timeInterval is a compulsory string'),
    });

    public docType?: string;
    public energyAmountMarketDocumentMrid: string; // PK
    public activationDocumentMrid: string; // FK1
    public registeredResourceMrid?: string;
    public quantity: string;
    public measurementUnitName: string;
    public revisionNumber?: string;
    public businessType?: string;
    public docStatus?: string;
    public processType?: string;
    public classificationType?: string;
    public areaDomain: string;
    public senderMarketParticipantMrid: string;
    public senderMarketParticipantRole: string;
    public receiverMarketParticipantMrid: string;
    public receiverMarketParticipantRole: string;
    public createdDateTime: string;
    public timeInterval: string;
}
